overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,373 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package devicestate implements the manager and state aspects responsible
+// for the device identity and policies.
+package devicestate
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+)
+
+// DeviceManager is responsible for managing the device identity and device
+// policies.
+type DeviceManager struct {
+	state      *state.State
+	keypairMgr asserts.KeypairManager
+	runner     *state.TaskRunner
+}
+
+// Manager returns a new device manager.
+func Manager(s *state.State) (*DeviceManager, error) {
+	runner := state.NewTaskRunner(s)
+
+	keypairMgr, err := asserts.OpenFSKeypairManager(dirs.SnapDeviceDir)
+	if err != nil {
+		return nil, err
+
+	}
+
+	m := &DeviceManager{state: s, keypairMgr: keypairMgr, runner: runner}
+
+	runner.AddHandler("generate-device-key", m.doGenerateDeviceKey, nil)
+	runner.AddHandler("request-serial", m.doRequestSerial, nil)
+
+	return m, nil
+}
+
+func (m *DeviceManager) ensureOperational() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+	// XXX: auth.Device/SetDevice should probably move to devicestate
+	// they are not quite just about auth and also we risk circular imports
+	// (auth will need to mediate bits from devicestate soon)
+	device, err := auth.Device(m.state)
+	if err != nil {
+		return err
+	}
+
+	if device.Serial != "" {
+		// serial is set, we are all set
+		return nil
+	}
+
+	if device.Brand == "" || device.Model == "" {
+		// need first-boot, loading of model assertion info
+		if release.OnClassic {
+			// TODO: are we going to have model assertions on classic or need will need to cheat here?
+			return nil
+		}
+		// cannot proceed yet, once first boot is done these will be set
+		// and we can pick up from there
+		return nil
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "become-operational" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// XXX: some of these will need to be split and use hooks
+	// retries might need to embrace more than one "task" then,
+	// need to be careful
+
+	genKey := m.state.NewTask("generate-device-key", i18n.G("Generate device key"))
+	requestSerial := m.state.NewTask("request-serial", i18n.G("Request device serial"))
+	requestSerial.WaitFor(genKey)
+
+	chg := m.state.NewChange("become-operational", i18n.G("Setting up device identity"))
```

Comment:
"Initialize device" (note the "ing" isn't typical for changes), let's also avoid calling out the identity here.

"device-setup" might be a better name instead of "become-operational" as well.